ext/opcache/jit/zend_jit_vm_helpers.c

```diff
@@ -1062,6 +1062,13 @@ zend_jit_trace_stop ZEND_FASTCALL zend_jit_trace_execute(zend_execute_data *ex,
 
 		trace_flags = ZEND_OP_TRACE_INFO(opline, offset)->trace_flags;
 		if (trace_flags) {
+			/* if inlined functions are too long, stop current tracing and restart a new one */
+			if (trace_buffer[idx-1].op == ZEND_JIT_TRACE_ENTER && idx > JIT_G(max_inline_func_length)) {
```

Comment:
This doesn't limit the inlined function length. This limits function inlining to the first trace instructions.

Reply:
> This doesn't limit the inlined function length. This limits function inlining to the first trace instructions.

Actually, JAVA JIT support inlined function up to a particular size. Here we limit size by separating the long trace by max_inline_func_length.

If we limit the inlined function length, some functions may never be JITTed.

Reply:
Your code allows inlining only if it occurs in the first 16 instructions of the trace. It easelly prohibits inlining of the short functions.

```
---- TRACE 44 start (side trace 43/3) WP_Scripts::localize() /home/www/bench/wordpress-3.6/wp-includes/class.wp-scripts.php:148
0027 FE_FREE V8 ; op1(array)
0028 T10 = ROPE_INIT 3 string("var ")
0029 T10 = ROPE_ADD 1 T10 CV1($object_name) ; op2(string)
0030 T9 = ROPE_END 2 T10 string(" = ")
0031 INIT_FCALL 1 96 string("json_encode")
     >init json_encode
0032 SEND_VAR CV2($l10n) 1 ; op1(array)
0033 V10 = DO_ICALL
     >call json_encode
0034 T8 = FAST_CONCAT T9 V10 ; op1(string) op2(string)
0035 CV6($script) = FAST_CONCAT T8 string(";") ; op1(string)
0036 T8 = ISSET_ISEMPTY_CV (empty) CV3($after) ; op1(undef)
0037 ;JMPNZ T8 0042
0042 INIT_METHOD_CALL 2 THIS string("get_data")
     >init WP_Dependencies::get_data
0043 SEND_VAR_EX CV0($handle) 1 ; op1(string)
0044 SEND_VAL_EX string("data") 2
0045 V8 = DO_FCALL
     >enter WP_Dependencies::get_data
---- TRACE 44 abort (JIT inlined function and skip current trace)
---- TRACE 44 start (enter) WP_Dependencies::get_data() /home/www/bench/wordpress-3.6/wp-includes/class.wp-dependencies.php:163
0002 T3 = FETCH_OBJ_IS THIS string("registered") ; val(array)
0003 T2 = ISSET_ISEMPTY_DIM_OBJ (isset) T3 CV0($handle) ; op1(array) op2(string) val(object)
0004 ;JMPNZ T2 0006
0006 T3 = FETCH_OBJ_IS THIS string("registered") ; val(array)
0007 T2 = FETCH_DIM_IS T3 CV0($handle) ; op1(array) op2(string) val(object)
0008 T3 = FETCH_OBJ_IS T2 string("extra") ; op1(object of class _WP_Dependency) val(array)
0009 T2 = ISSET_ISEMPTY_DIM_OBJ (isset) T3 CV1($key) ; op1(array) op2(string) val(undef)
0010 ;JMPNZ T2 0012
0011 RETURN bool(false)
---- TRACE 44 stop (return)
```

JAVA doesn't use tracing JIT. The current PHP tracing rules were mainly developed on top of LuaJIT.